storage/innobase/trx/trx0trx.cc

```diff
@@ -942,13 +938,11 @@ trx_t::assign_temp_rseg()
 	compile_time_assert(ut_is_2pow(TRX_SYS_N_RSEGS));
 
 	/* Choose a temporary rollback segment between 0 and 127
-	in a round-robin fashion. Because rseg_slot is not protected by
-	atomics or any mutex, race conditions are possible, meaning that
-	multiple transactions that start modifications concurrently
-	will write their undo log to the same rollback segment. */
-	static ulong	rseg_slot;
+	in a round-robin fashion. */
+	static std::atomic<ulong>	rseg_slot;
 	trx_rseg_t*	rseg = trx_sys.temp_rsegs[
-		rseg_slot++ & (TRX_SYS_N_RSEGS - 1)];
+		rseg_slot.fetch_add(1, std::memory_order_relaxed)
+		& (TRX_SYS_N_RSEGS - 1)];
```

Comment:
Likewise, here I would only change the variable declaration:
```c++
static Atomic_counter<unsigned> rseg_slot;
```
Note: I believe that 32 bits is enough here. I would avoid the `long` and `ulong` types because they are only 32 bits wide on 64-bit Windows.

Reply:
done.